File: 0437_pathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root: return 0
        self.path = []
        self.count = 0
        self.dfs(root, targetSum)
        return self.count

    def dfs(self, node, targetSum):
        if not node:
            return
        for i in range(len(self.path)):
            self.path[i] -= node.val
            if self.path[i] == 0: self.count += 1
        if targetSum - node.val == 0: self.count += 1
        self.path.append(targetSum - node.val)
        self.dfs(node.left, targetSum)
        self.dfs(node.right, targetSum)
        for i in range(len(self.path) - 1):
            self.path[i] += node.val
        self.path.pop()

# Alternative solution

class Solution1:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root: return 0
        self.result = 0
        self.dfs(root, targetSum, 0, collections.defaultdict(int, {0:1}))
        return self.result

    def dfs(self, node, targetSum, currentSum, memo):
        if not node: return
        currentSum += node.val
        previousSum = currentSum - targetSum
        self.result += memo[previousSum]
        memo[currentSum] += 1
        self.dfs(node.left, targetSum, currentSum, memo)
        self.dfs(node.right, targetSum, currentSum, memo)
        memo[currentSum] -= 1
        
# Alternative solution

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.total = 0
        self.dfs(root, targetSum)
        return self.total

    def dfs(self, node, target):
        if not node:
            return {}
        leftDict = self.dfs(node.left, target)
        rightDict = self.dfs(node.right, target)
        if node.val == target:
            self.total += 1
        newDict = {node.val: 1}
        for num in leftDict.keys():
            newNum = num + node.val
            newDict[newNum] = newDict.get(newNum, 0) + leftDict[num]
            if newNum == target:
                self.total += leftDict[num]
        for num in rightDict.keys():
            newNum = num + node.val
            newDict[newNum] = newDict.get(newNum, 0) + rightDict[num]
            if newNum == target:
                self.total += rightDict[num]
        return newDict

File: 0001-0500/0437_pathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int count = 0;

    public int pathSum(TreeNode root, int targetSum) {
        dfs(root, (long)targetSum);
        return count;
    }

    public HashMap<Long, Integer> dfs(TreeNode root, long targetSum) {
        if (root == null) return new HashMap<>();

        HashMap<Long, Integer> left = dfs(root.left, targetSum);
        HashMap<Long, Integer> right = dfs(root.right, targetSum);
        HashMap<Long, Integer> total = new HashMap<>();

        total.put((long)root.val, total.getOrDefault((long)root.val, 0) + 1);
        count += total.getOrDefault(targetSum, 0);

        left.forEach((key, val) -> {
            long newVal = key + root.val;
            if (newVal == targetSum) count += val;
            total.put(newVal, total.getOrDefault(newVal, 0) + val);         
        });

        right.forEach((key, val) -> {
            long newVal = key + root.val;
            if (newVal == targetSum) count += val;
            total.put(newVal, total.getOrDefault(newVal, 0) + val);
        });

        return total;
    }
}

// Alternative solution

class Solution {
    public int count = 0;

    public int pathSum(TreeNode root, int targetSum) {
        HashMap<Long, Integer> memo = new HashMap<>(){{put(0L, 1);}};
        dfs(root, targetSum, memo, 0L);
        return count;
    }

    public void dfs(TreeNode root, int targetSum, HashMap<Long, Integer> memo, long currentSum) {
        if (root == null) {
            return;
        }

        currentSum += (long) root.val;
        count += memo.getOrDefault(currentSum - targetSum, 0);
        memo.put(currentSum, memo.getOrDefault(currentSum, 0) + 1);
        dfs(root.left, targetSum, memo, currentSum);
        dfs(root.right, targetSum, memo, currentSum);
        memo.put(currentSum, memo.getOrDefault(currentSum, 0) - 1);
    }
}
